repl/repl.go

```diff
@@ -57,8 +61,121 @@ func Start(in io.Reader, out io.Writer) {
 		prompt.OptionPrefix("⧐  "),
 		prompt.OptionLivePrefix(changeLivePrefix),
 		prompt.OptionTitle("abs-repl"),
+		prompt.OptionHistory(getHistory()),
 	)
 	p.Run()
+	// we get here on ^D from the prompt
+	saveHistory()
+}
+
+/*
+Support for abs history file in the interactive REPL:
+
+1) The current ABS_HISTORY_FILE is loaded into the prompt.Run() cycle
+   using prompt.OptionHistory(getHistory()). This also loads the local history as well.
+   Default ABS_HISTORY_FILE is "~/.abs_history".
+2) Append each non-null, unique next line passed from prompt to the executor() to the local history.
+   NB. the live prompt history shows duplicate next lines, but they are not saved to the local history.
+3) Save the local history whenever the prompt.Run() loop exits (^D) or the executor() exits (on quit).
+   Write the local history to the ABS_HISTORY_FILE up to ABS_MAX_HISTORY_LINES (default 1000 lines).
+
+Note that ABS_HISTORY_FILE and ABS_MAX_HISTORY_LINES variables may come from the OS environment.
+*/
+
+const (
+	ABS_HISTORY_FILE      = "~/.abs_history"
+	ABS_MAX_HISTORY_LINES = "1000"
+)
+
+var (
+	historyFilePath string   // fully expanded ABS_HISTORY_FILE path
+	historyMaxLines int      // ABS_MAX_HISTORY_LINES
+	history         []string // local history
+)
+
+// expand full path to ABS_HISTORY_FILE for current user and get ABS_MAX_HISTORY_LINES
+func expandHistoryFilePath() string {
```

Comment:
this function both modifies the `historyMaxLines` "global" var as well as returns the path -- but from the name one could not deduce it's doing the former. Maybe you could just have a function called `getHistoryConfiguration(...) (max_lines, path)`? Then you could, in the `init` function of the `repl` module:

```
func init() {
  ...
  ...
  path, maxLines = getHistoryConfiguration()
  history = getHistory(path, maxLines)
}
```

and in the `Start` function you could simply:

```
prompt.OptionHistory(history),
```

In this way you rely on 2 less global vars (history vs path + max + history) and it'd feel kind of "cleaner" to me.

What do you think?

Reply:
Ok, can do.